src/test/java/com/redhat/rhjmc/containerjfr/tui/tcp/SocketClientReaderWriterTest.java

```diff
@@ -0,0 +1,120 @@
+package com.redhat.rhjmc.containerjfr.tui.tcp;
+
+import com.redhat.rhjmc.containerjfr.core.log.Logger;
+import org.hamcrest.MatcherAssert;
+import org.hamcrest.Matchers;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import java.io.IOException;
+import java.io.OutputStreamWriter;
+import java.net.Socket;
+import java.util.Scanner;
+import java.util.concurrent.Semaphore;
+
+import static org.mockito.Mockito.*;
+
+@ExtendWith(MockitoExtension.class)
+class SocketClientReaderWriterTest {
+    @Mock
+    Logger logger;
+
+    @Mock
+    Semaphore semaphore;
+
+    @Mock
+    Socket s;
+
+    @Mock
+    Scanner scanner;
+
+    @Mock
+    OutputStreamWriter writer;
+
+    SocketClientReaderWriter scrw;
+
+    @BeforeEach
+    void setup() throws Exception {
+        scrw = new SocketClientReaderWriter(logger, semaphore, s, scanner, writer);
+    }
+
+    @Test
+    void shouldClose() throws Exception {
+        scrw.close();
+
+        verify(scanner).close();
+        verify(writer).close();
+        verify(s).close();
+    }
+
+    @Test
+    void shouldReadLine() throws Exception {
+        String result = "read line result";
+        when(scanner.nextLine()).thenReturn(result);
+
+        reset(semaphore);
```

Comment:
What's the purpose of this reset?

Reply:
On `L63` and `L64`, there are verifications for `semaphore` to be properly released for exactly once. Since there is already a `.release()` call in the `SocketClientReaderWriter::new`, which is called in `setup()`, `semaphore` needs to be reset or there would be two `.release()` calls detected at the time of verification.

Reply:
That's what I figured. I'm not sure if I would rather have the `reset` or just `verify(semaphore, times(2)).whatever()`, but in either case there should be a comment explaining that there's an implicit/hidden `release` at construction time. Or perhaps there should just be a single `reset` done in the initial test setup?

Reply:
Or perhaps I should ask why the testing-only constructor needs to perform this initial `release` on the semaphore to begin with - since it's testing-only we can/probably should assume that the passed in instance will be a mock, so what does the first release achieve?

Reply:
I'd prefer a `reset()` as the last statement in `setup()` and comment there. Will fix in the next commit.

Reply:
In the non-testing constructor, `semaphore` is created with `0` initial permits. The `.release()` call in that constructor is critical or else there would be a deadlock. I thought the test constructor should follow the same pattern. But you are right, I now come to a realization it is indeed unnecessary as the passed in `semaphore` is mocked anyway.

I'll delete this initial `.release()` call in the testing constructor, as well as all `.reset()`s.